### PR TITLE
GitHub Actions .github/workflows/30min-iiab-test-install-raspios.yml for CI/CD & QA 

### PR DESCRIPTION
**Pull request by neomatrixcode**
* add .github/workflows/10min-iiab-test-install-raspbian.yml

### CODE REVIEW

**Comment by neomatrixcode** [comment]
add a virtual machine for testing Raspberry Pi Zero 2 W Arm Cortex-A53 (64 bits)

**Comment by holta** [comment]
This is incredible!  A few quick questions before merging:

- This PR suggests that IIAB should probably (any objections?) change the way it derives the value for `rpi_model` (this is IIAB's litmus test indicating whether or not it's running on a Raspberry Pi).  At the moment the var is set to "none" during such QEMU tests emulating RPi, which seems wrong!?  So if such a change is desired (i.e. to make these RPi simulations more legit) we'd revise the 2 lines below — to instead use the emulated Raspberry Pi Model extracted from (the last line of) `/proc/cpuinfo`:

  https://github.com/iiab/iiab/blob/58961b862c203d3152338b6e7e3e0849c0b55b38/scripts/local_facts.fact#L116-L117

  Given that [GitHub Actions](https://github.com/iiab/iiab/actions/runs/4987409883/jobs/8929198895?pr=3579#step:3:332) shows...

  image

- Does anyone know why line numbers are sometimes messy when using https://github.com/pguyot/arm-runner-action ?
  
  image

  Whatever's causing this... might also be related to the reason llne numbers sometimeso change completely, when one reloads test results pages like https://github.com/iiab/iiab/actions/runs/4987409883/jobs/8929198895?pr=3579 ?

- Maybe rename this PR's file in .github/workflows to something like `30min-iiab-test-install-raspios.yml` ?

  And if so, likewise rename the human-readable job name (string at the top line of the file) ?

- Does anyone know if emulating Raspberry Pi 4 or Raspberry Pi 400 is be just as slow (about 30min in total, as a result of the overhead from QEMU's emulation?)

Related:

- PR #3569

File: .github/workflows/10min-iiab-test-install-raspbian.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+          cpu_info: cpuinfo/raspberrypi_zero2_w_arm64
+          base_image: raspios_lite_arm64:latest
+    steps:
+      #- run: echo "🎉 The job was automatically triggered by a ${{ github.event_name }} event."
+      #- run: echo "🔎 The name of your branch is ${{ github.ref }} and your repository is ${{ github.repository }}."
+      #- name: Check out repository code
+      - uses: actions/checkout@v2
+      - uses: pguyot/arm-runner-action@v2
+        with:
+          image_additional_mb: 1024
```
Comment (holta) [inline .github/workflows/10min-iiab-test-install-raspbian.yml]: Doubling Raspberry Pi "Zero 2 W" RAM from 512MB to 1GB during emulation is of course cheating (!) but perhaps wise here for now 😅

**Comment by neomatrixcode** [comment]
1.- agree to use /proc/cpuinfo
2.- I had not realized that, I will look for why it happens
3.- Changing the name to 30min-iiab-test-install-raspios.yml looks fine to me
4.- I don't know, I'll do a time test of QEMU in windows 10

**Comment by holta** [comment]
My opinion: This is good enough to merge, and a very helpful, basic, automated ARM smoke-test!

But let's give it a few hours in case @jvonau, @tim-moody, @georgejhunt, @deldesir have any final suggestions on Tuesday?

**Comment by holta** [comment]
> 1.- agree to use /proc/cpuinfo

Done!  Hopefully this helps make QEMU's ARM tests substantially more meaningful:

- PR #3580

  image

  https://github.com/iiab/iiab/actions/runs/4988356082/jobs/8931008679?pr=3579#step:3:625

**Comment by tim-moody** [comment]
my only question is whether local vars none is a complete enough test, and would ansible-playbook <playbook.yml> --syntax-check do as much?

I'm not sure how to do it, but if the triggering mechanism knew which role was affected and used runrole to test that particular role it might save time and focus the test.

**Comment by holta** [comment]
> my only question is whether local vars none is a complete enough test,

Good question.  We might possibly ramp things up to use [local_vars_small.yml](https://github.com/iiab/iiab/blob/master/vars/local_vars_small.yml) later, which is the equivalent to default_vars.yml.

> and would ansible-playbook <playbook.yml> --syntax-check do as much?

It's possible we should add the above too ?

In addition to these GitHub Actions smoke tests — which do substantially more than syntax checking — a few examples are: pip failures, apt failures, download failures, poorly thought out code that doesn't complete, etc.

> I'm not sure how to do it, but if the triggering mechanism knew which role was affected and used runrole to test that particular role it might save time and focus the test.

Excellent question.  QA test automation is challenging.  Ideally that will happen in future!

**Comment by holta** [comment]
> ansible-playbook <playbook.yml> --syntax-check

The above would increase test coverage quite substantially, at very little cost, providing a relatively quick response to PR authors...

If it can be made to check all PR's 🤞

(In coming days we should probably add this rapid spot check of iiab-stages.yml — e.g. right after scripts/ansible installs ansible-core.  Doing this in the [ubuntu-latest workflow](https://github.com/iiab/iiab/blob/master/.github/workflows/10min-iiab-test-install.yml) is probably sufficient?)

https://github.com/iiab/iiab/blob/0cc3a064a9ab9135d5321a520ea3ed05bec95ad0/iiab-install#L5-L6

**Comment by tim-moody** [comment]
btw it looks like - uses: actions/checkout@v3 is now v3

**Comment by holta** [comment]
> btw it looks like - uses: actions/checkout@v3 is now v3

@neomatrixcode might you have time to check that out?

At least for the [ubuntu-latest workflow](https://github.com/iiab/iiab/blob/master/.github/workflows/10min-iiab-test-install.yml)  ?

**Comment by neomatrixcode** [comment]
ok, not problem

**Comment by holta** [comment]
> ok, not problem

If you can also get @tim-moody's `--syntax-check` working with the ubuntu-latest workflow — based on the [Fail Fast](https://en.wikipedia.org/wiki/Fail-fast)  principle, that would be a huge win too?!

With something like the following?

```
ansible-playbook -i ansible_hosts iiab-stages.yml --syntax-check
```

Ansible Tips:

- https://github.com/iiab/iiab/wiki/Technical-Contributors-Guide#female_detective-understanding-ansible
- https://github.com/iiab/iiab/blob/0cc3a064a9ab9135d5321a520ea3ed05bec95ad0/iiab-install#L197-L201
- https://github.com/iiab/iiab/blob/0cc3a064a9ab9135d5321a520ea3ed05bec95ad0/test.yml#L1-L2

**Comment by holta** [comment]
@neomatrixcode I'm tempted to merge this PR as is?

(Or LMK your preference, and don't hesitate to create other PR's — if you have other great ideas helping us become more responsive to IIAB communities!)

**Comment by neomatrixcode** [comment]
I don't plan to make any more changes.

**Comment by tim-moody** [comment]
git diff --name-status might be used to determine roles affected

**Comment by neomatrixcode** [comment]
> Does anyone know if emulating Raspberry Pi 4 or Raspberry Pi 400 is be just as slow (about 30min in total, as a result of the overhead from QEMU's emulation?)

I have done tests in windows 10, and indeed the ARM-x86_64 emulation is slow

https://unix.stackexchange.com/questions/675310/why-qemu-is-extremly-slow-how-to-fast-it-up

**Comment by jvonau** [comment]
> If you can also get @tim-moody's `--syntax-check` working with the ubuntu-latest workflow — based on the [Fail Fast](https://en.wikipedia.org/wiki/Fail-fast) principle, that would be a huge win too?!

@holta should practice what he preaches by leading by example or get of the soapbox.

**Comment by holta** [comment]
> > ansible-playbook <playbook.yml> --syntax-check
> 
> The above would increase test coverage quite substantially

Oops maybe not:

> I was looking for the same, but was not satisfied by the `--syntax-check` option, since it **does not work its way down to the roles.** A more complete check can be performed with [ansible-lint](https://docs.ansible.com/ansible-lint/)

- https://ansible-lint.readthedocs.io
- https://stackoverflow.com/questions/35339512/ansible-fast-way-to-check-syntax/56775736#56775736

**Comment by holta** [comment]
> > Does anyone know if emulating Raspberry Pi 4 or Raspberry Pi 400 is be just as slow (about 30min in total, as a result of the overhead from QEMU's emulation?)
> 
> I have done tests in windows 10, and indeed the ARM-x86_64 emulation is slow
> 
> https://unix.stackexchange.com/questions/675310/why-qemu-is-extremly-slow-how-to-fast-it-up

Will QEMU 8.0.0 (https://qemu.org/2023/04/20/qemu-8-0-0/) be substantially faster than QEMU 7.x for ARM emulation — and if so hopefully GitHub Actions with https://github.com/pguyot/arm-runner-action will take advantage this year or soon?!
